### ISSUE TITLE
Invalid range causes loop to be skipped.

### ISSUE THREAD

**Issue by arp2600** [rust-lang/rust#51692]
In a for loop with a range, if the range value is assigned to a type to small for the range, the loop will be skipped.

I tried this code:
```
fn main() {
    let mut counter: u64 = 0;
    for i in 0..256 {
        println!("i = {}", i);
        let x: u8 = i;
        counter += 1;
    }
    
    println!("looped {} times", counter);
}
```
Expected output:
```
i = 0
i = 1
...
i = 255
looped 256 times
```
Actual output:
```
looped 0 times
```

A warning is given that the literal is out of range, but I would have expected this to be a compilation error.
```
warning: literal out of range for u8
 --> src/main.rs:4:17
  |
4 |     for i in 0..256 {
  |                 ^^^
  |
  = note: #[warn(overflowing_literals)] on by default
```

Commenting out the line `let x: u8 = i;` causes the code to run correctly.

I tried this on stable and nightly with debug and release.
Below I've included a link to the rust playground demonstrating the issue.
https://play.rust-lang.org/?gist=36c8b0e6a621a419311cbfec1ce9cdcb&version=undefined&mode=undefined

**Comment by oli-obk** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/47213
